egs/wsj/s5/steps/nnet3/chain/train.py

```diff
@@ -0,0 +1,707 @@
+        """.format(command = run_opts.command,
+                   dir = dir, iter = iter, next_iter = iter + 1,
+                   shrink = shrinkage_value, best_model_index =  best_model))
+
+    try:
+        for i in range(1, num_jobs + 1):
+            os.remove("{0}/{1}.{2}.raw".format(dir, iter + 1, i))
+    except OSError:
+        raise Exception("Error while trying to delete the raw models")
+
+    new_model = "{0}/{1}.mdl".format(dir, iter + 1)
+
+    if not os.path.isfile(new_model):
+        raise Exception("Could not find {0}, at the end of iteration {1}".format(new_model, iter))
+    elif os.stat(new_model).st_size == 0:
+        raise Exception("{0} has size 0. Something went wrong in iteration {1}".format(new_model, iter))
+
+def CheckForRequiredFiles(feat_dir, tree_dir, lat_dir):
+    for file in ['{0}/feats.scp'.format(feat_dir), '{0}/ali.1.gz'.format(tree_dir),
+                 '{0}/final.mdl'.format(tree_dir), '{0}/tree'.format(tree_dir),
+                 '{0}/lat.1.gz'.format(lat_dir), '{0}/final.mdl'.format(lat_dir),
+                 '{0}/num_jobs'.format(lat_dir), '{0}/splice_opts'.format(lat_dir)]:
+        if not os.path.isfile(file):
+            raise Exception('Expected {0} to exist.'.format(file))
+
+# args is a Namespace with the required parameters
+def Train(args, run_opts):
+    arg_string = pprint.pformat(vars(args))
+    logger.info("Arguments for the experiment\n{0}".format(arg_string))
+
+    # Check files
+    CheckForRequiredFiles(args.feat_dir, args.tree_dir, args.lat_dir)
+
+    # Set some variables.
+    num_jobs = train_lib.GetNumberOfJobs(args.tree_dir)
+    feat_dim = train_lib.GetFeatDim(args.feat_dir)
+    ivector_dim = train_lib.GetIvectorDim(args.online_ivector_dir)
+
+    # split the training data into parts for individual jobs
+    # we will use the same number of jobs as that used for alignment
+    train_lib.SplitData(args.feat_dir, num_jobs)
+    shutil.copy('{0}/tree'.format(args.tree_dir), args.dir)
+    f = open('{0}/num_jobs'.format(args.dir), 'w')
+    f.write(str(num_jobs))
+    f.close()
+
+    config_dir = '{0}/configs'.format(args.dir)
+    var_file = '{0}/vars'.format(config_dir)
+
+    [model_left_context, model_right_context, num_hidden_layers] = train_lib.ParseModelConfigVarsFile(var_file)
+    # Initialize as "raw" nnet, prior to training the LDA-like preconditioning
+    # matrix.  This first config just does any initial splicing that we do;
+    # we do this as it's a convenient way to get the stats for the 'lda-like'
+    # transform.
+    if (args.stage <= -6):
+        logger.info("Creating phone language-model")
+        chain_lib.CreatePhoneLm(args.dir, args.tree_dir, run_opts, lm_opts = args.lm_opts)
+
+    if (args.stage <= -5):
+        logger.info("Creating denominator FST")
+        chain_lib.CreateDenominatorFst(args.dir, args.tree_dir, run_opts)
+
+    num_leaves = chain_lib.GetNumberOfLeaves(args.tree_dir)
```

Comment:
I removed the GetNum* call. But I use the GetNum* function in make_configs.py scripts to get the number of leaves from the tree dir, so I am keeping the function.